Add Linux implementation

pablojimpas opened the pull request:
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)

Feature: Add Linux support using [geoclue.dart](https://github.com/canonical/geoclue.dart). Based on previous work by @jpnurmi.

### :arrow_heading_down: What is the current behavior?

There is no Linux implementation.

### :new: What is the new behavior (if this is a feature change)?

The following methods of the platform interface are implemented:

  - checkPermission -> There's no permissions API
  - isLocationServiceEnabled -> TODO: Check if GeoClue D-Bus service is running.
  - getLastKnownPosition -> I'm not sure if GeoClue provides this exact functionality.
  - getCurrentPosition
  - getLocationAccuracy
  - getPositionStream

### :boom: Does this PR introduce a breaking change?

Maybe? Users of this package might be expecting an exception being thrown on Linux for unimplemented methods and now it's no longer the case. 

### :bug: Recommendations for testing

  - Linux system with D-Bus daemon running.
  - Make sure you have the GeoClue D-Bus service running.

There are some tests but I've also tested this manually on my system both with the example app that @jpnurmi created and with a real app overriding the dependencies and it's good enough so far for my use case.

### :memo: Links to relevant issues/docs

Closes #1008 

### :thinking: Checklist before submitting

- [x] I made sure all projects build.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning).
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I followed the style guide lines ([code style guide](https://github.com/Baseflow/flutter-geolocator/blob/master/CONTRIBUTING.md)).
- [x] I updated the relevant documentation.
- [x] I rebased onto current `master`.

jpnurmi commented:
Awesome, thank you! Sorry for dropping the ball.

pablojimpas commented:
> Awesome, thank you! Sorry for dropping the ball.

Thanks to you for your work! I've just taken what you did and added what I needed to make it work in my app. Hopefully more people find this useful 🙂

mvanbeusekom reviewed (changes requested):
Hi @pablojimpas and @jpnurmi,

Thank you for all your efforts in getting this integrated. Really appreciate it. The PR looks really promising I just had a few small remarks:

1. Please revert the changes to the `geolocator/pubspec.yaml` file. I think it would be a good idea to not directly fully endorse Linux support, especially since there are some `TODO`s still left in the code.
2. In the light of the above I made a suggestion on updating the `geolocator_linux/README.md` and explain how developers can add Linux support to their apps.
3. There is a `TODO` remark in  `geolocator_linux/lib/src/geolocator_linux.dart` that is not directly clear to me, my question would be if you could add a bit more detailed explanation to it.
4. Would you be so kind to add the CI configuration so GitHub actions will also validate the geolocator_linux packages? To do so please take the following steps:
    a. Create a file called `geolocator_linux.yaml` in the location `.github/workflows`;
    b. Add the following contents:
```yaml
name: geolocator_linux

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches: [ main ]
    paths:
    - 'geolocator_linux/**'
  pull_request:
    branches: [ main ]
    paths:
    - 'geolocator_linux/**'

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  build:
    name: Linux platform package

    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    env:
      source-directory: ./geolocator_linux
      example-directory: ./geolocator_linux/example 

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v2
      
      # Make sure the stable version of Flutter is available
      - uses: subosito/flutter-action@v1
        with:
          channel: 'stable'

      # Download all Flutter packages
      - name: Download dependencies
        run: flutter pub get
        working-directory: ${{env.source-directory}}

      # Run Flutter Format to ensure formatting is valid
      - name: Run Flutter Format
        run: flutter format --set-exit-if-changed .
        working-directory: ${{env.source-directory}}

      # Run Flutter Analyzer
      - name: Run Flutter Analyzer
        run: flutter analyze
        working-directory: ${{env.source-directory}}

      # Build Linux version of the example App
      - name: Run Linux build
        run: flutter build linux --release
        working-directory: ${{env.example-directory}}

      # Run all unit-tests with code coverage
      - name: Run unit tests
        run: flutter test --coverage
        working-directory: ${{env.source-directory}}

      # Upload code coverage information
      - uses: codecov/codecov-action@v1
        with:
          file: ${{env.source-directory}}/coverage/lcov.info # optional
          name: geolocator_linux (Linux Platform Package) # optional
          fail_ci_if_error: true
   ```

pablojimpas commented on geolocator_linux/lib/src/geolocator_linux.dart:
```diff
@@ -0,0 +1,72 @@
+
+class GeolocatorLinux extends GeolocatorPlatform {
+  static Future<void> registerWith() async {
+    // TODO: Platform.environment['XDG_CURRENT_DESKTOP'];
```
@jpnurmi wrote that, I assume that's the intention, the `GeolocatorGnome` class has a more concrete implementation for some methods, since the GNOME desktop environment provides a convenient way of checking the service status. I think the idea is to support different desktop environments were they do things slightly different, I'm not sure how KDE Plasma does this. However, I'm not using either desktop environment and the basic functionality works fine with the generic `GeolocatorLinux` class.

jpnurmi replied:
Sorry for such a vague comment. Yes, it's about adding support for other desktop environments. I noticed that the platform interface had API to open location settings etc. where it gets very desktop-specific.

mvanbeusekom replied:
Thanks @jpnurmi for the explanation, would it make sense to update the method to the following implementation? This way we at least catch the gnome desktop and default to the `GeolocatorLinux` implementation on all other desktops.

```dart
  static Future<void> registerWith() async {
    String? currentDesktop = Platform.environment['XDG_CURRENT_DESKTOP'];

    if (currentDesktop == null) {
      GeolocatorPlatform.instance = GeolocatorLinux(GeoClueManager());
      return;
    }

    switch (currentDesktop) {
      case 'GNOME':
        GeolocatorPlatform.instance = GeolocatorGnome(GeoClueManager());
        break;
      default:
        GeolocatorPlatform.instance = GeolocatorLinux(GeoClueManager());
    }
  }
```

This would show others how future desktop implementations could be added.

jpnurmi replied:
Be careful with `XDG_CURRENT_DESKTOP` because it's not as simple as it seems. :) For example, it's "ubuntu:GNOME" on my Ubuntu 22.04 desktop, and VS Code blatantly overrides it to "Unity" when running apps straight from VS Code.

jpnurmi replied:
The latter is a VS Code bug because it sets the value for its own purposes and then leaks it, but the bug reports will arrive here if the handling is too strict. The way I had envisioned this was to use the GNOME implementation by default because even KDE/Qt-based desktops have GNOME/GTK settings schemas installed, but it's certainly possible to set up a more exotic system without them so I guess handling the value is safer. Alternatively, `GeolocatorGnome` could be adjusted to catch exceptions to gracefully handle potentially missing settings.

mvanbeusekom replied:
I have updated the sample suggestion to at least catch the `ubuntu:GNOME` case.

pablojimpas replied:
Perfect! I've updated the method with this structure in case we need to handle more cases in the future.

pablojimpas commented on geolocator/pubspec.yaml:
```diff
@@ -21,6 +21,8 @@ flutter:
       windows:
         default_package: geolocator_windows
+      linux:
+        default_package: geolocator_linux
```
I've just seen your comment on the initial [windows support](https://github.com/Baseflow/flutter-geolocator/pull/954#pullrequestreview-860192192), maybe we can proceed in a similar way here.

I've just reverted all the changes in the `geolocator` package itself and clarified the `geolocator_linux` README. Once we're more confident with the implementation we can add the `geolocator_linux` dependency to `geolocator`, for now I think it's good enough to release a `0.0.1` version so that more people can start integrating it with their apps.

pablojimpas commented:
Hi @mvanbeusekom, thanks for your valuable comments. I've addressed most of your concerns but we may need input from @jpnurmi on what's the idea with the `XDG_CURRENT_DESKTOP` platform checking long-term.

Otherwise, I think this is ready for a `0.0.1` version.

mvanbeusekom reviewed (commented):
Thanks for the updates. Could you check this [comment](https://github.com/Baseflow/flutter-geolocator/pull/1049#discussion_r863656053) and see if it makes sense?

I feel it would be better to have a structure in place where we fallback on the `GeolocatorLinux` class when GNOME is not available. However I have very limited Linux experience so if this doesn't make any sense feel free to let me know ;).

mvanbeusekom reviewed (approved):
LGTM, thanks @pablojimpas, @jpnurmi for all your work on this!

jpnurmi commented:
Thanks @mvanbeusekom and @pablojimpas for finishing it off!

mvanbeusekom commented:
P.S. @pablojimpas, @jpnurmi, I would like to announce the release of Linux support via Twitter. Do you mind me mentioning your name and if you are on Twitter is it ok to mention your Twitter handle (could you please provide your Twitter handle :P)?

jpnurmi commented:
Sounds good, let's put the word out. Mentioning is fine by me :) I'm @jpnurmi on Twitter, too.

pablojimpas commented:
> P.S. @pablojimpas, @jpnurmi, I would like to announce the release of Linux support via Twitter. Do you mind me mentioning your name and if you are on Twitter is it ok to mention your Twitter handle (could you please provide your Twitter handle :P)?

Awesome! It's okey to me, I don't have a Twitter account nowadays but feel free to link my GitHub account if you want.
